providers/adfs_test.go

```diff
@@ -0,0 +1,147 @@
+package providers
+
+import (
+	"context"
+	"encoding/base64"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+
+	"github.com/coreos/go-oidc"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+type fakeAdfsJwks struct{}
+
+func (fakeAdfsJwks) VerifySignature(_ context.Context, jwt string) (payload []byte, err error) {
+	decodeString, err := base64.RawURLEncoding.DecodeString(strings.Split(jwt, ".")[1])
+	if err != nil {
+		return nil, err
+	}
+	return decodeString, nil
+}
+
+func testAdfsProvider(hostname string) *ADFSProvider {
+
+	o := oidc.NewVerifier(
+		"https://issuer.example.com",
+		fakeAdfsJwks{},
+		&oidc.Config{ClientID: "https://test.myapp.com"},
+	)
+
+	p := NewADFSProvider(NewOIDCProvider(
+		&ProviderData{
+			ProviderName: "",
+			LoginURL:     &url.URL{},
+			RedeemURL:    &url.URL{},
+			ProfileURL:   &url.URL{},
+			ValidateURL:  &url.URL{},
+			Scope:        "",
+			Verifier:     o,
+		}))
+
+	if hostname != "" {
+		updateURL(p.Data().LoginURL, hostname)
+		updateURL(p.Data().RedeemURL, hostname)
+		updateURL(p.Data().ProfileURL, hostname)
+		updateURL(p.Data().ValidateURL, hostname)
+	}
+
+	return p
+}
+
+func testAdfsBackend() *httptest.Server {
+
+	authResponse := `
+		{
+			"access_token": "my_access_token",
+			"id_token": "my_id_token",
+			"refresh_token": "my_refresh_token" 
+		 }
+	`
+	userInfo := `
+		{
+			"email": "samiracho@email.com"
+		}
+	`
+
+	refreshResponse := `{ "access_token": "new_some_access_token", "refresh_token": "new_some_refresh_token", "expires_in": "32693148245", "id_token": "new_some_id_token" }`
+
+	authHeader := "Bearer adfs_access_token"
+
+	return httptest.NewServer(http.HandlerFunc(
```

Comment:
I recently wrote this for a hackday project if you want to give it a whirl and see how it works out: https://github.com/oauth2-proxy/mockoidc

I'm going to finalize it and include it in our OIDCProvider tests & new integration tests on `oauthproxy.go` soon as well.